src/testing/cluster/storage_checker.zig

```diff
@@ -86,59 +87,115 @@ pub const StorageChecker = struct {
-            log.debug("{}: replica_checkpoint: checkpoint={} area={s} value={?x:0>32}", .{
+            log.debug("{}: {s}: checkpoint={} area={s} value={?x:0>32}", .{
                 superblock.replica_index.?,
+                caller,
                 replica_checkpoint_op,
                 @tagName(area),
                 checkpoint_actual.get(area),
             });
         }
 
-        const checkpoint_expect = checker.checkpoints.get(replica_checkpoint_op) orelse {
-            if (!syncing) {
-                // This replica is the first to reach op_checkpoint.
-                // Save its state for other replicas to check themselves against.
-                var checkpoint: Checkpoint = undefined;
-                for (std.enums.values(CheckpointArea)) |area| {
-                    checkpoint.set(area, checkpoint_actual.get(area).?);
+        if (checker.checkpoints.get(replica_checkpoint_op)) |checkpoint_expect| {
+            var mismatch: bool = false;
+            for (std.enums.values(CheckpointArea)) |area| {
+                const checksum_actual = checkpoint_actual.get(area) orelse continue;
+                const checksum_expect = checkpoint_expect.get(area);
+                if (checksum_expect != checksum_actual) {
+                    log.warn("{}: {s}: mismatch " ++
+                        "area={s} expect={x:0>32} actual={x:0>32}", .{
+                        superblock.replica_index.?,
+                        caller,
+                        @tagName(area),
+                        checksum_expect,
+                        checksum_actual,
+                    });
+
+                    mismatch = true;
                 }
-                try checker.checkpoints.putNoClobber(replica_checkpoint_op, checkpoint);
             }
-            return;
-        };
-
-        var mismatch: bool = false;
-        for (std.enums.values(CheckpointArea)) |area| {
-            const checksum_actual = checkpoint_actual.get(area) orelse continue;
-            const checksum_expect = checkpoint_expect.get(area);
-            if (checksum_expect != checksum_actual) {
-                log.warn("{}: replica_checkpoint: mismatch " ++
-                    "area={s} expect={x:0>32} actual={x:0>32}", .{
-                    superblock.replica_index.?,
-                    @tagName(area),
-                    checksum_expect,
-                    checksum_actual,
-                });
-
-                mismatch = true;
+            if (mismatch) return error.StorageMismatch;
+        } else {
+            // This replica is the first to reach op_checkpoint.
+            // Save its state for other replicas to check themselves against.
+            var checkpoint: Checkpoint = undefined;
+            for (std.enums.values(CheckpointArea)) |area| {
+                if (checkpoint_actual.get(area)) |area_checksum| {
+                    checkpoint.set(area, area_checksum);
+                } else {
+                    return;
```

Comment:
Hm, this `return` looks suspicions --- if a replica has a partial checkpoint, and is the first one with a checkpoint, we wont' be veryfying it. If this is OK, the comment would be helpful. 

Although even with a comment, it feels like something which could easily mutate to skipping all checkpoints. Can we add some sanity checks that `checker.checkpoints` eventually have all of the checkpoints?

Reply:
> Can we add some sanity checks that checker.checkpoints eventually have all of the checkpoints?

Hmm yeah this is confusing + possibly a footgun. I'm going to rework the storage checker so that it incrementally adds whatever info it can, even if (for example) it first encounters the checkpoint from a syncing replica.

(Context:
As it is now, it is possible (albeit unlikely) for checkpoints to be entirely missed from the storage checker's point of view.
The reason is that if a replica crashes part-way through checkpointing (for example, after it has written 3 out of 4 superblock copies), and then the replica recovers (into the new 3/4 checkpoint), then it never invokes the "on checkpoint" hook.)